pkgs/by-name/ap/apkeditor/package.nix

```diff
@@ -0,0 +1,164 @@
+      self = stdenv.mkDerivation (
+        {
+          __darwinAllowLocalNetworking = true;
+
+          buildInputs = [ jre ];
+          nativeBuildInputs = [ gradle ];
+
+          gradleFlags = [ "-Dfile.encoding=utf-8" ];
+          gradleBuildTask = "jar";
+          doCheck = true;
+
+          inherit outJar;
+          installPhase = ''
+            runHook preInstall
+            install -Dm644 build/libs/*.jar $out/${outJar}
+            runHook postInstall
+          '';
+        }
+        // args
+        // {
+          meta.sourceProvenance = with lib.sourceTypes; [
+            fromSource
+            binaryBytecode # mitm cache
+          ];
+        }
+      );
+    in
+    self;
+
+  arsclib = REAndroidLibrary {
+    pname = "arsclib";
+    version = "1.3.5-unstable-2024-10-21"; # 1.3.5 is not new enough for APKEditor
+    projectName = "ARSCLib";
+
+    src = fetchFromGitHub {
+      owner = "REAndroid";
+      repo = "ARSCLib";
+      rev = "ed6ccf00e56d7cce13e8648ad46a2678a6093248";
+      hash = "sha256-jzd7xkc4O+P9hlGsFGGl2P3pqVvV5+mDyKTRUuGfFSA=";
+    };
+
+    mitmCache = gradle.fetchDeps {
+      pkg = arsclib;
+      data = ./deps-arsclib.json;
+    };
+  };
+
+  smali = REAndroidLibrary {
+    pname = "smali";
+    version = "2024-10-15";
+    projectName = "smali";
+
+    src = fetchFromGitHub {
+      owner = "REAndroid";
+      repo = "smali-lib";
+      rev = "c781eafb31f526abce9fdf406ce2c925fec20d28";
+      hash = "sha256-6tkvikgWMUcKwzsgbfpxlB6NZBAlZtTE34M3qPQw7Y4=";
+    };
+
+    mitmCache = gradle.fetchDeps {
+      pkg = smali;
+      data = ./deps-smali.json;
+    };
+    gradleBuildTask = "build";
+
+    installPhase = ''
+      runHook preInstall
+      install -Dm644 smali/build/libs/*-fat.jar $out/${smali.outJar}
+      runHook postInstall
+    '';
+  };
+
+  jcommand = REAndroidLibrary {
+    pname = "jcommand";
+    version = "2024-09-20";
+    projectName = "JCommand";
+
+    src = fetchFromGitHub {
+      owner = "REAndroid";
+      repo = "JCommand";
+      rev = "714b6263c28dabb34adc858951cf4bc60d6c3fed";
+      hash = "sha256-6Em+1ddUkZBCYWs88qtfeGnxISZchFrHgDL8fsgZoQg=";
+    };
+
+    mitmCache = gradle.fetchDeps {
+      pkg = jcommand;
+      data = ./deps-jcommand.json;
+    };
+  };
+
+  apkeditor =
+    let
+      pname = "apkeditor";
+      version = "1.4.1";
+      projectName = "APKEditor";
+    in
+    REAndroidLibrary {
+      inherit pname version projectName;
+
+      # When you need to update deps-*.json for the dependencies (e.g. for smali),
+      # uncomment the following line and run `nix build apkeditor.smali.mitmCache.updateScript`.
+
+      # inherit arsclib smali jcommand;
+
+      src = fetchFromGitHub {
+        owner = "REAndroid";
+        repo = "APKEditor";
+        rev = "V${version}";
+        hash = "sha256-a72j9qGjJXnTFeqLez2rhBSArFVYCX+Xs7NQd8CY5Yk=";
+      };
+
+      nativeBuildInputs = [
+        gradle
+        arsclib
+        smali
+        jcommand
```

Comment:
If anything, these would go into buildInputs but does gradle even pick these up from here? You're linking them into some sort of classpath dir later manually.

Reply:
The gradle script uses the binary libraries with hard coded path relative to the project dir. After finishing building the jar, those files are not needed anymore because the resultant jar can be run standalone.

Reply:
I'm not sure you replied to the correct comment here.

Reply:
```suggestion
```

Reply:
I thought they have to be in `nativeBuildInputs` to tell nix that they needs to be built before building `apkeditor` and that they are not needed once `apkeditor` is already built.

Reply:
Do you mean that as long as `${arsclib}` is in a string, `arsclib` will be built? So, if I use `${lib.getExe makeWrapper}`, I don't need `makeWrapper` in `nativeBuildInputs` either?

Reply:
Yup. Nix only cares about references. Every path that is referenced in the drv must be realised before the drv can be realised.

Once a drv is realised, the actual runtime dependencies are based on what is referenced in the resulting output path. It just scans for string references; this is pretty basic and simple but works surprisingly well.

The actual purpose of *buildInputs is to make the packages available in search paths such as `$PATH` for executables. If the build system takes a library out of the pkg-config search path and links it in the binary, that's the references that will draw the library in as a runtime dependency. 

While you could reference the exe directly like that, it's better to use nativeBuildInputs for executables wherever possible in order to get the packages automagically spliced as the correct platform config which is important for cross-compilation. You can enforce this in non-cross scenarios too using `strictDeps = true;` which you should basically always set. If you must interpolate a package or executable thereof directly, always take the package form `buildPackages`. That's not relevant here though.